src/Storage/TSDatabaseView.m

```diff
@@ -265,11 +266,10 @@ + (void)asyncRegisterSecondaryDevicesDatabaseView
 + (NSDate *)localTimeReceiveDateForInteraction:(TSInteraction *)interaction {
     NSDate *interactionDate = interaction.date;
 
-    if ([interaction isKindOfClass:[TSIncomingMessage class]]) {
-        TSIncomingMessage *message = (TSIncomingMessage *)interaction;
-
-        if (message.receivedAt) {
-            interactionDate = message.receivedAt;
+    if ([interaction isKindOfClass:[TSMessage class]]) {
```

Comment:
> How are we going to sort existing messages by bestReceivedAt if it's nil?

1. We try to use `bestReceivedAt`. If its not set (because this is a message received before these changes)...
2. We try to use `bestReceived`.  Legacy messages only set that for incoming messages so if its not set...
3. We fallback to `date`.

Previously we did steps 2 & 3.  We're adding 1.

> The default implementation of `bestReceivedAtDate` could return `interaction.date` if `receivedAt` is nil.

I've taken your suggestions and a) DRY'd up the "default to TSInteraction.date" logic b) renamed `bestReceivedAtDate` to `receiptDateForSorting`. c) Pulled up `receiptDateForSorting` into `TSInteraction`.

> As is, it might work for now, but I suspect the next time we bump the version on this view we'll lose all sorting for the old messages.

I don't understand why?